code/modules/logging/log_holder.dm

```diff
@@ -14,51 +11,143 @@ GLOBAL_PROTECT(logger)
+			continue
+
+		if(!initial(category_type.schema_version))
+			stack_trace("log category type '[category_type]' does not have a valid schema version, skipping")
+			continue
+
+		var/master_category = initial(category_type.master_category)
+		if(master_category)
+			sub_categories[master_category] += list(category_type)
+			continue
+		category_tree[category_type] = list()
+
+	// Sort the sub categories
+	for(var/datum/log_category/master as anything in sub_categories)
+		if(!(master in category_tree))
+			stack_trace("log category [master] is an invalid master category as its a sub category")
+			continue
+		for(var/datum/log_category/sub_category as anything in sub_categories[master])
+			if(initial(sub_category.secret) != initial(master.secret))
+				stack_trace("log category [sub_category] has a secret status that differs from its master category [master]")
+			category_tree[master] += list(sub_category)
+
+	return category_tree.Copy()
+
+/// Initializes the given log category and populates the list of contained categories based on the sub category list
+/datum/log_holder/proc/init_log_category(datum/log_category/category_type, list/datum/log_category/sub_categories)
+	var/datum/log_category/category_instance = new category_type
+
+	var/list/contained_categories = list()
+	for(var/datum/log_category/sub_category as anything in sub_categories)
+		sub_category = new sub_category
+		sub_category.master_category = category_instance
+		log_categories[sub_category.category] = sub_category
+		contained_categories[sub_category.category] = sub_category.schema_version
+
+	log_categories[category_instance.category] = category_instance
+	contained_categories[category_instance.category] = category_instance.schema_version
+
+	var/list/category_header = list(
+		LOG_HEADER_INIT_TIMESTAMP = big_number_to_text(logging_start_timestamp),
+		LOG_HEADER_ROUND_ID = big_number_to_text(GLOB.round_id),
+		LOG_HEADER_SECRET = category_instance.secret,
+		LOG_HEADER_SCHEMA_LIST = contained_categories,
+		LOG_HEADER_CATEGORY = category_instance.category,
+	)
+	rustg_file_write("[json_encode(category_header)]\n", category_instance.get_output_file(null))
+
+
+/// Adds an entry to the given category, if the category is disabled it will not be logged.
+/// If the category does not exist, we will CRASH and log to the error category.
+/// the data list is optional and will be recursively json serialized.
 /datum/log_holder/proc/Log(category, message, list/data)
-	if(!initialized || shutdown)
-		CRASH("Attempted to perform logging before initializion or after shutdown!")
+	// This is Log because log is a byond internal proc
+	if(shutdown)
+		CRASH("Attempted to perform logging after shutdown!")
+
+	if(!istext(message))
+		CRASH("Attempted to log a non-text message! [message]")
+
+	if(!category)
+		CRASH("Attempted to log to a null category! [message]")
+
+	if(data && !islist(data))
+		CRASH("Attempted to log a non-list data! [data]")
```

Comment:
None of these are allowed.

logging has to be very liberal in what it accepts and must try as hard as it can to log.

also you can't have logging messages going to the runtime log given they may contain secret logs.

for after shutdown, it still has to log to disk immediately.

for non-text messages, it should stringify the message and log. (with a warning)

for non-category messages, it should default to the LOG_CATEGORY_NOT_FOUND category but then prepend the category to the start message (incase the category includes the message)

for non-list data, it should either stringify the data and append it to the end the message or turn data into `list("data" = data)`